drivers/eeprom/eeprom_at2x.c

```diff
@@ -662,3 +662,107 @@ DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C64_A125_BUS_API &st_m24c64_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C64_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24c32_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24c32_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C32_A125_BUS_API &st_m24c32_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C32_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24c16_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24c16_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C16_A125_BUS_API &st_m24c16_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C16_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24c08_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24c08_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C08_A125_BUS_API &st_m24c08_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C08_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24c04_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24c04_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C04_A125_BUS_API &st_m24c04_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C04_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24c02_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24c02_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24C02_A125_BUS_API &st_m24c02_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24C02_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
```

Comment:
This is a lot of effort to duplicate common code.  There should be a better option.

If the problem is to use the right `DT_DRV_COMPAT`, note that it's possible to define multiple compatibles on a node.
For instance:
```
	compatible = "st,m24c64-a125", "atmel,at2x";
```
(dt spec mentions this should go from the most specific to the most generic).
Then , you can use `#define DT_DRV_COMPAT atmel_at2x`

Reply:
Thanks to the input of @pdgendt the code duplication is now down to almost none:

    #define DT_DRV_COMPAT atmel_at25
    DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_SPI_BUS,
                      EEPROM_AT2X_BUS_API_SPI, ASSERT_AT25_ADDR_W_VALID)
    #undef DT_DRV_COMPAT